### PR TITLE
Add type in PropertyKey and enforce boolean type

### PR DESCRIPTION
**Pull request by rongrong**
### What changes are proposed in this pull request?

Each PropertyKey / PropertyKey.Template should have a type. Property key values
will be validated against the type when set.

This PR only adds boolean type, all non-boolean property keys right now have
type String. There are some boolean properties that are depending on other
property values, thus default is a string in the form of "${...}". So right
now String is always allowed for property key values. Ideally, property key
dependency should be handled separately from simple values.

### Why are the changes needed?

Code safety and efficiency

### CODE REVIEW

File: core/common/src/main/java/alluxio/conf/InstancedConfiguration.java [hunk 1]
```diff
@@ -453,15 +467,29 @@ private String lookupRecursively(String base, Set<String> seen)
...
         throw new UnresolvablePropertyException(ExceptionMessage
             .UNDEFINED_CONFIGURATION_KEY.getMessage(match));
       }
-      resolved = resolved.replaceFirst(REGEX_STRING, Matcher.quoteReplacement(value));
+      resolved = resolved.replaceFirst(REGEX_STRING, Matcher.quoteReplacement(stringValue));
+    }
+    if (key != null) {
+      resolvedValue = key.parseValue(resolved);
+    }
+    else {
```
Comment (apc999) [inline core/common/src/main/java/alluxio/conf/InstancedConfiguration.java:489]: nit: no line break before `else` as code convention in this repo.
We basically follow Google Java style guide: https://google.github.io/styleguide/javaguide.html#s4.1.2-blocks-k-r-style

```suggestion
    } else {
```

File: core/common/src/main/java/alluxio/conf/InstancedConfiguration.java [hunk 2]
```diff
@@ -579,8 +607,8 @@ void checkTieredStorage() {
     int globalTiers = getInt(PropertyKey.MASTER_TIERED_STORE_GLOBAL_LEVELS);
     Set<String> globalTierAliasSet = new HashSet<>();
     for (int i = 0; i < globalTiers; i++) {
-      globalTierAliasSet
-          .add(get(PropertyKey.Template.MASTER_TIERED_STORE_GLOBAL_LEVEL_ALIAS.format(i)));
+      globalTierAliasSet.add(
+              getString(PropertyKey.Template.MASTER_TIERED_STORE_GLOBAL_LEVEL_ALIAS.format(i)));
```
Comment (apc999) [inline core/common/src/main/java/alluxio/conf/InstancedConfiguration.java:611]: nit: we follow Google Java style of indentation: 

https://google.github.io/styleguide/javaguide.html#s4.5.2-line-wrapping-indent

File: core/common/src/main/java/alluxio/conf/PropertyKey.java [hunk 3]
```diff
@@ -4598,7 +4655,8 @@ public String toString() {
           .setDescription("The total memory overhead for bloom filters used for tracking")
           .setConsistencyCheckLevel(ConsistencyCheckLevel.WARN).setScope(Scope.CLIENT).build();
   public static final PropertyKey USER_CLIENT_CACHE_SHADOW_BLOOMFILTER_NUM =
-      new Builder(Name.USER_CLIENT_CACHE_SHADOW_BLOOMFILTER_NUM).setDefaultValue(4)
+      intBuilder(Name.USER_CLIENT_CACHE_SHADOW_BLOOMFILTER_NUM)
+              .setDefaultValue(4)
```
Comment (apc999) [inline core/common/src/main/java/alluxio/conf/PropertyKey.java]:
```suggestion
          .setDefaultValue(4)
```

File: core/common/src/main/java/alluxio/conf/PropertyKey.java [hunk 4]
```diff
@@ -7717,13 +7791,15 @@ public static PropertyKey fromString(String input) {
    * @param displayType how the property value should be displayed
    * @param isBuiltIn whether this is an Alluxio built-in property
    */
-  private PropertyKey(String name, String description, DefaultSupplier defaultSupplier,
-      String[] aliases, boolean ignoredSiteProperty, boolean isHidden,
-      ConsistencyCheckLevel consistencyCheckLevel, Scope scope, DisplayType displayType,
-      boolean isBuiltIn, boolean dynamic, Function<Object, Boolean> valueValidationFunction) {
+  private PropertyKey(String name, String description, Class<?> javaType,
```
Comment (apc999) [inline core/common/src/main/java/alluxio/conf/PropertyKey.java]: nit: javadoc `javaType`

File: core/common/src/main/java/alluxio/conf/PropertyKey.java [hunk 5]
```diff
@@ -7931,12 +8023,32 @@ public DisplayType getDisplayType() {
    * @return whether the value is a valid value of the property key
    */
   public boolean validateValue(Object value) {
+    if (!(value instanceof String)
+            && (mJavaType.equals(Boolean.class) || mJavaType.equals(Integer.class))
+            && !value.getClass().equals(mJavaType)) {
```
Comment (apc999) [inline core/common/src/main/java/alluxio/conf/PropertyKey.java]:
```suggestion
        && (mJavaType.equals(Boolean.class) || mJavaType.equals(Integer.class))
        && !value.getClass().equals(mJavaType)) {
```

File: core/common/src/main/java/alluxio/conf/PropertyKey.java [hunk 6]
```diff
@@ -7931,12 +8023,32 @@ public DisplayType getDisplayType() {
...
+      return false;
+    }
     if (mValueValidationFunction == null) {
       return true;
     }
     return mValueValidationFunction.apply(value);
   }
 
+  /**
+   * Parse value from string.
```
Comment (apc999) [inline core/common/src/main/java/alluxio/conf/PropertyKey.java]: nit
```suggestion
   * Parses value from string.
```

File: core/common/src/main/java/alluxio/util/FormatUtils.java [hunk 7]
```diff
@@ -249,23 +249,6 @@ public static long parseTimeSize(String timeSize) {
...
 
-  /**
-   * Parses a boolean from a string.
-   *
-   * @param bool the boolean to parse
-   * @return whether the value was "true" or "false"
-   * @throws IllegalArgumentException if the value cannot be parsed
-   */
-  public static boolean parseBoolean(String bool) throws IllegalArgumentException {
-    if (bool.equalsIgnoreCase("true")) {
```
Comment (apc999) [inline core/common/src/main/java/alluxio/util/FormatUtils.java:260]: question: if in the `alluxio-site.properties` file, a user sets a property to be `True` or `FALSE` (allowed previously),
will the current implementation throw RTE?

**Review by apc999 — COMMENTED** [review]
some minor comments.

btw, do you plan to add type like `Bytes` and `Ms` in the future?

**Comment by rongrong** [comment]
> some minor comments.
> 
> btw, do you plan to add type like `Bytes` and `Ms` in the future?
 
Yes, that's the plan. I was thinking whether it's better to use enum or java class for types. Using java class means that we will probably convert "ms" to duration and bytes to datasize. Using enum means we can keep using long. What's your thought?

File: core/common/src/main/java/alluxio/util/FormatUtils.java (hunk 7 above)
Comment (rongrong) [reply core/common/src/main/java/alluxio/util/FormatUtils.java:260]: Do we have a test testing this scenario?

Comment (rongrong) [reply core/common/src/main/java/alluxio/util/FormatUtils.java:260]: `Boolean.parseBoolean` would return true matching true ignore cases and return false for all other values. The difference is here is whether you want to enforce `false` to only match false ignore cases. So all existing cases should still work, but "no" and "yes" will also not throw in the future.

**Pull request merged**